src/utils/TodoContext.tsx

```diff
@@ -0,0 +1,205 @@
+  error: ErrorMessage.NO_ERRORS,
+  setError: () => {},
+  isLoading: false,
+  setIsLoading: () => {},
+  tempTodo: null,
+  setTempTodo: () => {},
+  loadingTodosIDs: [],
+  setLoadingTodosIDs: () => {},
+});
+
+type Props = {
+  children: React.ReactNode;
+};
+
+export const TodosProvider: FC<Props> = ({ children }) => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [filter, setFilter] = useState<Filter>(Filter.ALL);
+  const [query, setQuery] = useState('');
+  const [error, setError] = useState<ErrorMessage>(ErrorMessage.NO_ERRORS);
+  const [isLoading, setIsLoading] = useState(false);
+  const [loadingTodosIDs, setLoadingTodosIDs] = useState<number[]>([]);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+
+  const addTodo = (newTodo: Omit<Todo, 'id'>) => {
+    setIsLoading(true);
+    setLoadingTodosIDs(prev => [...prev, 0]);
+
+    sendTodoToServer(newTodo)
+      .then(response => {
+        setTodos(prevTodos => [...prevTodos, response]);
+        setQuery('');
+      })
+      .catch(() => {
+        setError(ErrorMessage.ADD_TODO_ERROR);
+      })
+      .finally(() => {
+        setTempTodo(null);
+        setIsLoading(false);
+        setLoadingTodosIDs(prev => prev.filter(todoId => todoId !== 0));
+      });
+  };
+
+  const removeTodo = (id: number) => {
+    setIsLoading(true);
+    setLoadingTodosIDs(prev => [...prev, id]);
+    deleteTodoFromServer(id)
+      .then(() => {
+        setTodos(prevTodos => prevTodos.filter(todo => todo.id !== id));
+      })
+      .catch(() => {
+        setError(ErrorMessage.DELETE_TODO_ERROR);
+      })
+      .finally(() => {
+        setIsLoading(false);
+        setLoadingTodosIDs(prev => prev.filter(todoId => todoId !== id));
+      });
+  };
+
+  const toggleTodoStatus = (id: number, status: boolean) => {
+    setIsLoading(true);
+    setLoadingTodosIDs(prev => [...prev, id]);
+
+    const todoToUpdate = todos.find(todo => todo.id === id);
+
+    if (todoToUpdate) {
+      toggleStatusTodo(id, status)
+        .then(() => {
+          setTodos(prevTodos =>
+            prevTodos.map(prev =>
+              prev.id === id ? { ...prev, completed: status } : prev,
+            ),
+          );
+        })
+        .catch(() => {
+          setError(ErrorMessage.UPDATE_TODO_ERROR);
+        })
+        .finally(() => {
+          setIsLoading(false);
+          setLoadingTodosIDs(prev => prev.filter(todoId => todoId !== id));
+        });
+    }
+  };
+
+  const toggleTodo = (id: number) => {
+    const todoToUpdate = todos.find(todo => todo.id === id);
+
+    if (todoToUpdate) {
+      toggleTodoStatus(id, !todoToUpdate.completed);
+    }
+  };
+
+  const toggleAllTodos = (status: boolean) => {
+    todos.forEach(todo => {
+      if (todo.completed !== status) {
+        toggleTodoStatus(todo.id, status);
+      }
+    });
+  };
+
+  const renameTodo = (
+    id: number,
+    newTitle: string,
+    setEditingTodoId: (id: number | null) => {},
```

Comment:
I do not really like this `setEditingTodoId` arg, because it is not related to `renameTodo` logic it's more about specific UI.

Can we return Promise from this function and then just handle it in our component? 

If it's not the solution, rename `setEditingTodoId` to something more generic like `callback`.